include/boost/numeric/ublas/eigen_solver.hpp

```diff
@@ -0,0 +1,395 @@
+* The output Eigenvalues as well as Eigenvectors can be both real or complex. 
+* \tparam M type of the objects stored in the vector. Expected to be a matrix (like matrix<double>)
+*/
+
+template <class M> 
+class eigen_solver {
+private:
+	M matrix;
+	M hessenberg_form;
+	M real_schur_form;
+	M transform_accumulations;
+	bool has_complex_part;
+	bool has_eigenvalues;
+	bool has_eigenvectors;
+	M eigenvalues_real;
+	M eigenvalues_complex;
+	M eigenvectors;
+	M eigenvectors_real;
+	M eigenvectors_complex;
+	eig_solver_params solver_params;
+public:
+	/// \brief Constructor that takes a matrix and an (optional) argument whether to extract eigenvectors also. 
+	/// The only argument that it needs is the matrix for which we wish to compute the Eigenvalues (and optionally Eigenvectors). The second option
+	/// specifies whether we wish to compute only the Eigenvalues (EIGVAL) or both Eigenvalues as well as Eigenvectors(EIGVEC). The default option 
+	/// is that it only computes the Eigenvalues. 
+	/// \param m Matrix for which we need to extract the Eigenvalues 
+	/// \param params Can have two values EIGVAL or EIGVEC.
+	BOOST_UBLAS_INLINE 
+	explicit eigen_solver(M &m, eig_solver_params params = EIGVAL) :
+	matrix(m), has_complex_part(false), solver_params(params)
+	{
+		has_eigenvalues = has_eigenvectors = false;
+		hessenberg_form = M(m);
+		compute(solver_params);
+	}
+
+	/// \brief Method to order the class to (re)compute the Eigenvalues and Eigenvectors.
+	/// Normally the user doesn't need to call this method explicitly as it called with the constructor. However, if initially the second argument
+	/// specified EIGVAL option and then the user wants to compute the Eigenvectors also, he will need to call this to recompute the 
+	/// Eigenvectors with the EIGVEC option.
+	/// \param params Can have two values EIGVAL or EIGVEC.
+	BOOST_UBLAS_INLINE 
+	void compute(eig_solver_params params = EIGVAL) {
+		if (params != solver_params) {
+			solver_params = params;
+		}
+
+		//If only eigen values are desired
+		if (params == EIGVAL) {
+			to_hessenberg<M>(hessenberg_form);
+			real_schur_form = M(hessenberg_form);
+			schur_decomposition<M>(real_schur_form);
+			extract_eigenvalues_from_schur();
+			has_eigenvalues = true;
+			has_eigenvectors = false;
+		}
+		else {
+			to_hessenberg<M>(hessenberg_form, transform_accumulations);
+			real_schur_form = M(hessenberg_form);
+			schur_decomposition<M>(real_schur_form, transform_accumulations);
+			extract_eigenvalues_from_schur();
+			extract_eigenvectors();
+			has_eigenvalues = true;
+			has_eigenvectors = true;
+		}
+
+
+	}
+
+	BOOST_UBLAS_INLINE
+	void extract_eigenvalues_from_schur() {
+		typedef typename M::size_type size_type;
+		typedef typename M::value_type value_type;
+
+		size_type n = real_schur_form.size1();
+		size_type i = size_type(0);
+
+		eigenvalues_real = zero_matrix<value_type>(n,n);
```

Comment:
The real and imaginary components of the eigenvalues should be returned as vectors, especially when eigenvectors are not required. Compare with [Eigen](https://github.com/eigenteam/eigen-git-mirror/blob/master/Eigen/src/Eigenvalues/EigenSolver.h).